sphinx/util/logging.py

```diff
@@ -449,10 +450,13 @@ def filter(self, record: SphinxWarningLogRecord) -> bool:  # type: ignore
         location = getattr(record, 'location', None)
         if isinstance(location, tuple):
             docname, lineno = location
-            if docname and lineno:
-                record.location = '%s:%s' % (self.app.env.doc2path(docname), lineno)
-            elif docname:
-                record.location = '%s' % self.app.env.doc2path(docname)
+            if docname:
+                if os.path.isabs(docname):
+                    record.location = docname
+                else:
+                    record.location = self.app.env.doc2path(docname)
+                if lineno:
+                    record.location += ":{}".format(lineno)
```

Comment:
Initially I've thought about fixing/handling this in doc2path - but I've figured that calling it with a path already is wrong, but have not checked other usages of it.

Reply:
Since https://github.com/sphinx-doc/sphinx/pull/7273 will be taken instead I've wondered if it would make sense to assert / and or warn that the location is always a docname / no full path here then?

Reply:
-1 for assert. Surely it would help Sphinx developers, but it brings trouble to users.
+0 for warning. It might help Sphinx developers (if they find warnings) and it does not harm users.

Reply:
Yeah, the "assert" should certainly not be used in production, but e.g. in tests.

Therefore a warning seems feasible: it could have a type that breaks tests, but otherwise gets only displayed / can be filtered.

Reply:
Some user uses sphinx via `python -m sphinx`. If my understanding correct, we should notice them to give `-O` option to invoke Sphinx, right?
It would be wonderful if it works only in testing.

Reply:
I think using `assert` is not really good, but will evaluate if a special warning class can be used, that would be not shown by default (but also would not hurt really I guess?), and cause an error during tests.

Will close this PR, and keep that in mind/inbox to follow up eventually.